File: c++/leetcode/230. Kth Smallest Element in a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  vector<int> arr;

public:
  vector<int> smallest(TreeNode *);
  int kthSmallest(TreeNode *root, int k)
  {
    return smallest(root).at(k - 1);
  }
};

vector<int> Solution::smallest(TreeNode *root)
{
  if (!root)
    return arr;
  smallest(root->left);
  arr.push_back(root->val);
  smallest(root->right);
  return arr;
}

File: python/leetcode/230. Kth Smallest Element in a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.kth = k
        self.res = 0

        self.inorder(root)
        return self.res

    def inorder(self, root):
        if root is None or self.kth == 0:
            return

        self.inorder(root.left)
        self.kth -= 1
        if self.kth == 0:
            self.res = root.val
        else:
            self.inorder(root.right)
